### ISSUE TITLE
Consider not warning on unobserved awaitable returned from generic methods

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#8988]
As [reported on SO](http://stackoverflow.com/q/35417106/41071), the following code causes CS4014 ("Because this call is not awaited, execution of the current method continues before the call is completed."):

``` c#
private Task _Task;

public async Task DoSomething()
{
    Interlocked.Exchange(ref _Task, null);
    …
}
```

As mentioned by Lucas Trzesniewski in the answer to the above linked SO question, this warning probably doesn't make sense when the method returns generic `T` that just happens to be a `Task` (or another awaitable).

Another example that supports this would be `Queue<Task>` where `Peek()` is used to get the top task and `Dequeue()` is used only to remove it, where the `Dequeue()` would also trigger this warning incorrectly.

My questions:
1. Is the assertion that methods that return a generic type that happens to be awaitable don't have to be awaited true? Do you have any counterexamples?
2. If the assertion is true, does it make sense to not show the warning in such cases?

**Comment by jmarolf** [comment]
This warning could be more descriptive.  Its caused by ignoring the return value of `Interlocked.Exchange`.  In your `Queue<Task>.Dequeue()` example, it might be reasonable to ignore the returned value, but how is the compiler to know?  It is generally bad practice to ignore returned values.  If you are certain that this warning is not applicable to you, then you can disable it with a pragma for that particular instance.
